src/VisualStudio/Core/Impl/SolutionExplorer/SourceGeneratedFileItems/SourceGeneratedFileItemSource.cs

```diff
@@ -0,0 +1,281 @@
+
+            // We must update the list on the UI thread, since the WPF elements bound to our list expect that
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            try
+            {
+                // We're going to incrementally update our items, ensuring we keep the object identity for things we didn't touch.
+                _items.BeginBulkOperation();
+
+                // Do we already have a "no files" placeholder item?
+                if (_items.Count == 1 && _items[0] is NoSourceGeneratedFilesPlaceholderItem)
+                {
+                    // We do -- if we have no items, we're done, since the placeholder is all that needs to be there;
+                    // otherwise remove it since we have real files now
+                    if (sourceGeneratedDocumentsForGeneratorById.Count == 0)
+                    {
+                        return;
+                    }
+                    else
+                    {
+                        _items.RemoveAt(0);
+                    }
+                }
+
+                for (int i = 0; i < _items.Count; i++)
+                {
+                    // If this item that we already have is still a generated document, we'll remove it from our list; the list when we're
+                    // done is going to have the new items remaining. If it no longer exists, remove it from list.
+                    if (!sourceGeneratedDocumentsForGeneratorById.Remove(((SourceGeneratedFileItem)_items[i]).DocumentId))
+                    {
+                        _items.RemoveAt(i);
+                        i--;
+                    }
+                }
+
+                // Whatever is left in sourceGeneratedDocumentsForGeneratorById we should add
+                if (sourceGeneratedDocumentsForGeneratorById.Count != 0)
+                {
+                    foreach (var document in sourceGeneratedDocumentsForGeneratorById.Values)
+                    {
+                        // Binary search to figure out where to insert
+                        int low = 0;
+                        int high = _items.Count;
+
+                        while (low < high)
+                        {
+                            int mid = (low + high) / 2;
+
+                            if (StringComparer.OrdinalIgnoreCase.Compare(document.HintName, ((SourceGeneratedFileItem)_items[mid]).HintName) < 0)
+                            {
+                                high = mid;
+                            }
+                            else
+                            {
+                                low = mid + 1;
+                            }
+                        }
+
+                        _items.Insert(low, new SourceGeneratedFileItem(document.Id, document.HintName, document.Project.Language, _workspace));
+                    }
+                }
+                else
```

Comment:
potentially make this early return instead of else for clarity

Reply:
Yeah, good idea since I did an early return earlier.